core/vm/instructions.go

```diff
@@ -496,6 +496,16 @@ func opExtCodeCopy(pc *uint64, evm *EVM, contract *Contract, memory *Memory, sta
 	return nil, nil
 }
 
+func opExtCodeHash(pc *uint64, evm *EVM, contract *Contract, memory *Memory, stack *Stack) ([]byte, error) {
+	slot := stack.peek()
+	if hash := evm.StateDB.GetCodeHash(common.BigToAddress(slot)); hash == (common.Hash{}) {
+		slot.SetUint64(0)
+	} else {
+		slot.SetBytes(hash.Bytes())
+	}
```

Comment:
Martin is right here, if the code *does* have a hash, you can return it. If there is no code however, 3 scenarios can happen:

 * The account is a plain account: the returned code hash should be Keccak256("") (ie. `c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470`)
 * The account is a precompile, this is an interesting case, but I think Keccak256("") would be the correct one.
 * The account does not exist: `00...0`

Please see https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1052.md#test-cases for details

Reply:
I think you could simplify this code via:

```go
slot := stack.peek()
slot.SetBytes(evm.StateDB.GetCodeHash(common.BigToAddress(slot)))
```

I.e. Your optimization of `slot.SetUint64(0)` is not worth it, because it entails the `hash == (common.Hash{})` comparison, which in theory should be equivalent runtime wise to `slot.SetBytes(hash.Bytes())`.